Source/JavaScriptCore/builtins/ModuleLoader.js

```diff
@@ -23,586 +23,642 @@
+function fetchModuleAndEvaluate(key, parameters, fetcher)
 {
     "use strict";
 
-    var importMap = @importMapStatus();
-    if (importMap)
-        await importMap;
-    var entry = await this.requestSatisfy(this.ensureRegistered(key), parameters, fetcher, new @Set);
-    return entry.key;
+    return this.fetchModule(key, parameters, fetcher).then(() => {
+        var entry = this.getModuleMapEntry(key, this.getModuleType(parameters));
+        @assert(!!entry);
+        return this.moduleEvaluation(entry, fetcher);
+    });
 }
 
 @visibility=PrivateRecursive
-function linkAndEvaluateModule(key, fetcher)
+function evaluateModule(key, fetcher)
 {
     "use strict";
 
-    var entry = this.ensureRegistered(key);
-    this.link(entry, fetcher);
+    var entry = this.moduleMap.@get(key);
+    @assert(!!entry);
     return this.moduleEvaluation(entry, fetcher);
 }
 
 @visibility=PrivateRecursive
-async function loadAndEvaluateModule(moduleName, parameters, fetcher)
+function loadModule(key, sourceCode, fetcher)
+{
+    "use strict";
+
+    var entry = this.ensureModuleMapEntry(key, @ModuleTypeJavaScript);
+    entry.fetchAndParsePromise ??= this.parseModule(key, sourceCode).then(moduleRecord => this.hostLoadImportedModule(entry, moduleRecord), error => {
+        entry.errorToRethrow = error;
+        return entry;
+    });
+
+    return @ensureImportMapsLoaded(() => {
+        return entry.fetchAndParsePromise
+            .then(entry => {
+                if (entry.errorToRethrow)
+                    throw entry.errorToRethrow;
+                return entry;
+            })
+            .then(() => this.fetchDescendantsOfAndLink(entry, fetcher));
+    });
+}
+
+@visibility=PrivateRecursive
+function loadModuleAndEvaluate(key, sourceCode, fetcher)
 {
     "use strict";
 
-    var importMap = @importMapStatus();
-    if (importMap)
-        await importMap;
-    var key = this.resolve(moduleName, @undefined, fetcher);
-    key = await this.loadModule(key, parameters, fetcher);
-    return await this.linkAndEvaluateModule(key, fetcher);
+    return this.loadModule(key, sourceCode, fetcher).then(() => {
+        var entry = this.moduleMap.@get(key);
+        @assert(!!entry);
+        return this.moduleEvaluation(entry, fetcher);
+    });
 }
 
 @visibility=PrivateRecursive
-async function requestImportModule(moduleName, referrer, parameters, fetcher)
+function requestImportModule(specifier, referrer, parameters, fetcher)
 {
     "use strict";
 
-    var importMap = @importMapStatus();
-    if (importMap)
-        await importMap;
-    var key = this.resolve(moduleName, referrer, fetcher);
-    var entry = await this.requestSatisfy(this.ensureRegistered(key), parameters, fetcher, new @Set);
-    await this.linkAndEvaluateModule(entry.key, fetcher);
-    return this.getModuleNamespaceObject(entry.module);
+    return @ensureImportMapsLoaded(() => {
+        var key = this.resolve(specifier, referrer, fetcher);
```

Comment:
Missing an exception catch as used in `innerModuleLoading`?

Reply:
Not actually: exception check in `innerModuleLoading` implements error propagation / caching from entry's requested modules to the entry itself (step 9.3 of https://html.spec.whatwg.org/multipage/webappapis.html#creating-a-javascript-module-script). In dynamic import, a failure to resolve an identifier ultimately returns a rejected promise (steps 8-9 of https://html.spec.whatwg.org/multipage/webappapis.html#hostloadimportedmodule).